### PR TITLE
Learn2code | Review of the love2code configs

### PR DESCRIPTION
**Pull request by revbucket**
Dirk, help?

### CODE REVIEW

File: src/olmo_core/nn/transformer/config.py [hunk 1]
```diff
@@ -821,3 +821,18 @@ def ngpt_like(
...
         )
+
+
+
+    @classmethod
+    def love2code_3B(cls, vocab_size: int, **kwargs) -> "TransformerConfig":
+        """ OLMo-core repro of GraniteCode 3B model """
+        return cls.llama_like(
+            d_model=2560,
+            hidden_size_multiplier=1.5,
```
Comment (dirkgr) [inline src/olmo_core/nn/transformer/config.py:962]: This isn't right. Based on the Granite paper, this should be `10240 // 2560`, but it also has to be a multiple of 128.

File: src/olmo_core/nn/transformer/config.py [hunk 2]
```diff
@@ -821,3 +821,18 @@ def ngpt_like(
...
+            vocab_size=vocab_size,
+            n_layers=32,
+            n_heads=32,
+            **kwargs
+        )
+        # num_params: 3607267840
```
Comment (dirkgr) [inline src/olmo_core/nn/transformer/config.py]: Where did you get this number?

File: src/scripts/train/learn2code/love2code-3B.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    NumpyDatasetConfig,
+    TokenizerConfig,
+    TokenizerName,
+)
+from olmo_core.train import TrainerConfig
+from olmo_core.train.callbacks import CheckpointerCallback, CometCallback, WandBCallback
+from typing import Callable, Dict, List, Tuple, Optional, cast
+import sys
+
+SEQUENCE_LENGTH = 2048
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: Short is OK for ablations? Also good for perf?

File: src/scripts/train/learn2code/love2code-3B.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    TokenizerConfig,
+    TokenizerName,
+)
+from olmo_core.train import TrainerConfig
+from olmo_core.train.callbacks import CheckpointerCallback, CometCallback, WandBCallback
+from typing import Callable, Dict, List, Tuple, Optional, cast
+import sys
+
+SEQUENCE_LENGTH = 2048 
+CHINCHILLA_5X_DURATION = Duration.tokens(36072678400 * 20 * 5)
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: Is that the real number of params? Should be non-embedding params. There might be a way to get this programmatically.

File: src/scripts/train/learn2code/love2code-3B.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+    dataset = NumpyDatasetConfig.from_data_mix(
+        Love2CodeDataMix.love2code_mix,
+        tokenizer=tokenizer,
+        mix_base_dir=root_dir,
+        sequence_length=SEQUENCE_LENGTH,
+        work_dir=get_work_dir(root_dir)
+        )
+    data_loader_config = NumpyDataLoaderConfig(
+        global_batch_size=global_batch_size, seed=34521, num_workers=4)
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: Guessing you need more data loading workers. A small model will run fast, so the data needs to come in faster. I'd start with 8 or 16.

File: src/scripts/train/learn2code/love2code-3B.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+def build_model_config(common: CommonComponents) -> TransformerConfig:
+    """
+    Note for DG/PW: Created a custom model (basically just copied what Pete sent me in slack)
+    """
+    return TransformerConfig.love2code_3B(
+        vocab_size=common.tokenizer.padded_vocab_size(),
+        compile=True,
+        dp_config=TransformerDataParallelConfig(
+            name=DataParallelType.hsdp, param_dtype=DType.bfloat16, reduce_dtype=DType.float32
+        ),
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py:125]: Set number of replicas to number of nodes.

File: src/scripts/train/learn2code/love2code-3B.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    del common
+    return AdamWConfig(
+        lr=4e-4,
+        weight_decay=0.1,
+        betas=(0.9, 0.95),
+        group_overrides=[
+            OptimGroupOverride(params=["embeddings.weight"], opts=dict(weight_decay=0.0))
+        ],
+        fused=True,
+    )
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py:145]: Does Granite have any guidelines here? 

LRs it still a black art. 7B was fine with up to `12e-4`. So I think you can go higher here. Let's do `12e-4` for this one as well. You're going to train these to the end before thinking about the results, right?

File: src/scripts/train/learn2code/love2code-3B.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+def build_trainer_config(common: CommonComponents) -> TrainerConfig:
+    """ Note for DG/PW: The only think I changed here was to add the max_duration in TrainerConfig """
+    return (
+        TrainerConfig(
+            save_folder=common.save_folder,
+            rank_microbatch_size=8 * 4096,
+            save_overwrite=True,
+            metrics_collect_interval=10,
+            cancel_check_interval=1,
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: Set to 10. Might not be necessary, but just in case.

File: src/scripts/train/learn2code/love2code-3B.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                ephemeral_save_interval=1000,
+                save_async=True,
+            ),
+        )
+        .with_callback(
+            "comet",
+            CometCallback(
+                name=common.run_name,
+                workspace="ai2",
+                project="OLMo-core-1B",
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: Put it into a different project?

File: src/scripts/train/learn2code/love2code-3B.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+                enabled=True,
+                cancel_check_interval=10,
+            ),
+        )
+        .with_callback(
+            "wandb",
+            WandBCallback(
+                name=common.run_name,
+                entity="ai2-llm",
+                project="OLMo-core-1B",
```
Comment (dirkgr) [inline src/scripts/train/learn2code/love2code-3B.py]: same

**Review by dirkgr — CHANGES_REQUESTED** [review]
Didn't get all the way though. Will do more later.

File: src/olmo_core/nn/transformer/config.py (hunk 1 above)
Comment (revbucket) [reply src/olmo_core/nn/transformer/config.py:962]: Au contraire. Setting hidden_size_multiplier to 1.5 actually means a 4x upscale for FFN because there's an implicit multiplication by 8/3 in here. Don't @me here, this is a pete decision

File: src/olmo_core/nn/transformer/config.py (hunk 2 above)
Comment (revbucket) [reply src/olmo_core/nn/transformer/config.py]: I built the model locally and asked it how many parameters it had

File: src/scripts/train/learn2code/love2code-3B.py (hunk 3 above)
Comment (revbucket) [reply src/scripts/train/learn2code/love2code-3B.py]: Copy from granite. We should go bigger, but trying to be faithful in the architecture repro

File: src/olmo_core/nn/transformer/config.py (hunk 2 above)
Comment (dirkgr) [reply src/olmo_core/nn/transformer/config.py]: Can you go fancy and do this calculation inside of this script?

**Comment by dirkgr** [comment]
Actually, the rest of the script was just boiler plate, so I guess I'm done after all.

File: src/scripts/train/learn2code/love2code-3B.py (hunk 7 above)
Comment (revbucket) [reply src/scripts/train/learn2code/love2code-3B.py:145]: > For the phase-1 pretraining, the learning rate follows a cosine schedule starting from 3 × 10−4 which decays to 3× 10−5 with an initial linear warmup step of 2k iterations.

From Granite. So I guess we jut replicate?

File: src/scripts/train/learn2code/love2code-3B.py (hunk 6 above)
Comment (revbucket) [reply src/scripts/train/learn2code/love2code-3B.py:125]: This may or may not be necessary. `num_replicas` seems to default to num_nodes

Asked Pete about it, so we'll see when he gets back

File: src/scripts/train/learn2code/love2code-3B.py (hunk 7 above)
Comment (revbucket) [reply src/scripts/train/learn2code/love2code-3B.py:145]: Okay, switched to 12e-4, linearly decaying to 12e-5, with 2k steps of warmup.

File: src/scripts/train/learn2code/love2code-3B.py (hunk 6 above)
Comment (dirkgr) [reply src/scripts/train/learn2code/love2code-3B.py:125]: You are 100% correct.
